File: iterableapi/src/main/java/com/iterable/iterableapi/IterableUtil.java
package com.iterable.iterableapi;

import android.content.Context;
import android.content.SharedPreferences;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;

import org.json.JSONObject;

import java.io.File;

public class IterableUtil {
    @VisibleForTesting
    static IterableUtilImpl instance = new IterableUtilImpl();

    private IterableUtil() { }

    static long currentTimeMillis() {
        return instance.currentTimeMillis();
    }

    public static String getAppVersion(Context context) {
        return instance.getAppVersion(context);
    }

    public static String getAppVersionCode(Context context) {
        return instance.getAppVersionCode(context);
    }

    /**
     * Maps a version string like "x.y.z" onto an integer like xxxyyyzzz
     * Example: "1.2.3" -> 1002003
     * @param versionString Version string
     * @return Integer representation of the version
     */
    static int convertVersionStringToInt(String versionString) {
        return instance.convertVersionStringToInt(versionString);
    }

    static void saveExpirableJsonObject(SharedPreferences preferences, String key, JSONObject object, long expirationInterval) {
        instance.saveExpirableJsonObject(preferences, key, object, expirationInterval);
    }

    static void saveExpirableValue(SharedPreferences preferences, String key, String value, long expirationInterval) {
        instance.saveExpirableValue(preferences, key, value, expirationInterval);
    }

    static String retrieveExpirableValue(SharedPreferences preferences, String key) {
        return instance.retrieveExpirableValue(preferences, key);
    }

    static JSONObject retrieveExpirableJsonObject(SharedPreferences preferences, String key) {
        return instance.retrieveExpirableJsonObject(preferences, key);
    }

    @Nullable
    static Long retrieveValidCampaignIdOrNull(final JSONObject json, final String key) {
        return instance.retrieveValidCampaignIdOrNull(json, key);
    }

    static boolean isValidCampaignId(final long campaignId) {
        return instance.isValidCampaignId(campaignId);
    }

    static File getSdkCacheDir(Context context) {
        return instance.getSdkCacheDir(context);
    }

    static File getSDKFilesDirectory(Context context) {
        return instance.getSDKFilesDirectory(context);
    }

    static File getDirectory(File folder, String subFolder) {
        return instance.getDirectory(folder, subFolder);
    }

    @Nullable
    static String readFile(File file) {
        return instance.readFile(file);
    }

    static boolean writeFile(File file, String content) {
        return instance.writeFile(file, content);
    }

    static boolean isUrlOpenAllowed(@NonNull String url) {
        return instance.isUrlOpenAllowed(url);
    }

    static String getWebViewBaseUrl() {
        return instance.getWebViewBaseUrl();
    }
}

File: iterableapi/src/main/java/com/iterable/iterableapi/IterableUtil.kt
package com.iterable.iterableapi

import android.content.Context
import android.content.SharedPreferences

import androidx.annotation.NonNull
import androidx.annotation.Nullable
import androidx.annotation.VisibleForTesting

import org.json.JSONObject

import java.io.File

object IterableUtil {
    @VisibleForTesting
    @JvmStatic
    internal var instance: IterableUtilImpl = IterableUtilImpl()

    @JvmStatic
    fun currentTimeMillis(): Long {
        return instance.currentTimeMillis()
    }

    @JvmStatic
    fun getAppVersion(context: Context): String {
        return instance.getAppVersion(context)
    }

    @JvmStatic
    fun getAppVersionCode(context: Context): String {
        return instance.getAppVersionCode(context)
    }

    /**
     * Maps a version string like "x.y.z" onto an integer like xxxyyyzzz
     * Example: "1.2.3" -> 1002003
     * @param versionString Version string
     * @return Integer representation of the version
     */
    @JvmStatic
    fun convertVersionStringToInt(versionString: String): Int {
        return instance.convertVersionStringToInt(versionString)
    }

    @JvmStatic
    fun saveExpirableJsonObject(preferences: SharedPreferences, key: String, obj: JSONObject, expirationInterval: Long) {
        instance.saveExpirableJsonObject(preferences, key, obj, expirationInterval)
    }

    @JvmStatic
    fun saveExpirableValue(preferences: SharedPreferences, key: String, value: String, expirationInterval: Long) {
        instance.saveExpirableValue(preferences, key, value, expirationInterval)
    }

    @JvmStatic
    fun retrieveExpirableValue(preferences: SharedPreferences, key: String): String? {
        return instance.retrieveExpirableValue(preferences, key)
    }

    @JvmStatic
    fun retrieveExpirableJsonObject(preferences: SharedPreferences, key: String): JSONObject? {
        return instance.retrieveExpirableJsonObject(preferences, key)
    }

    @Nullable
    @JvmStatic
    fun retrieveValidCampaignIdOrNull(json: JSONObject, key: String): Long? {
        return instance.retrieveValidCampaignIdOrNull(json, key)
    }

    @JvmStatic
    fun isValidCampaignId(campaignId: Long): Boolean {
        return instance.isValidCampaignId(campaignId)
    }

    @JvmStatic
    fun getSdkCacheDir(context: Context): File {
        return instance.getSdkCacheDir(context)
    }

    @JvmStatic
    fun getSDKFilesDirectory(context: Context): File {
        return instance.getSDKFilesDirectory(context)
    }

    @JvmStatic
    fun getDirectory(folder: File, subFolder: String): File {
        return instance.getDirectory(folder, subFolder)
    }

    @Nullable
    @JvmStatic
    fun readFile(file: File): String? {
        return instance.readFile(file)
    }

    @JvmStatic
    fun writeFile(file: File, content: String): Boolean {
        return instance.writeFile(file, content)
    }

    @JvmStatic
    fun isUrlOpenAllowed(@NonNull url: String): Boolean {
        return instance.isUrlOpenAllowed(url)
    }
}
